Feat/2 주문서생성

SerenityZenDev opened the pull request:
1. Order 엔티티 생성
2. OrderDetail 엔티티 생성
3. Order CRUD 레이어 생성
4. OrderDetail CRUD 레이어 생성
5. Order와 OrderDetail 연관관계 형성
6. Audit 작동여부 확인
7. 주문서 생성 기능 구현
8. Redis 장바구니와의 연동

Postman

주문 생성 성공

![주문생성_성공](https://github.com/user-attachments/assets/d20c98b5-0149-4ae3-a82c-51c8a9a0d75b)

주문 생성 실패 

Null Validation
![주문생성_실패_null](https://github.com/user-attachments/assets/5ee3e79d-f109-46d2-b86e-7462589ae95a)

UUID Validation
![주문생성_실패_UUID](https://github.com/user-attachments/assets/d3721a3e-6483-4236-a9cb-5d35d76b14a0)

장바구니에 내용물 없음
![주문생성_실패_장바구니](https://github.com/user-attachments/assets/f422b767-2da6-4c70-a239-da507eb7c16f)

SerenityZenDev committed:
feat: 주문서 생성 구현 (#2)
- Order 엔티티 생성
- Order CRUD 레이어 생성
- OrderDetail 엔티티 생성
- OrderDetail CRUD 레이어 생성
- 주문서 생성 기능 구현

SerenityZenDev committed:
refactor: 주문서 생성 구현 (#2)
- Order와 OrderDetail 사이 누락된 연관관계 구현
- Order와 OrderDetail 사이 순환참조로 인한 return값 조정을 위한 임시 DTO 적용

SerenityZenDev committed:
Merge branch 'dev' of https://github.com/BAEDALPING/order-service into feat/2-주문서생성

SerenityZenDev committed:
refactor: 주문서 생성 구현 (#2)
- Audit 적용
- dto 세부명칭 구분
- 공통 response 반환 적용

yooyouny commented on src/main/java/com/baedalping/delivery/order/dto/OrderCreateRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+import com.baedalping.delivery.order.entity.OrderDetail;
+import java.util.List;
+
+public class OrderCreateRequestDto {
```
dto에서 entity 레벨의 객체를 쓰는건 지양해야하지 않을까 싶은데 의견이 어떠실까요? 
요소가 같더라도 요청 데이터를 전송하는 역할을 하는 객체를 따로 만들어야 entity가 변경되더라도 dto에 영향을 주지 않을것 같은데요

SerenityZenDev replied:
해당 의견에는 동의하며 추후 수정 예정입니다.

다만 현재 주문서 생성을 위해 자동으로 가져올 수 있는 내용
{ex) 주문 상세 내역에 들어갈 상품 정보 등} 이 redis 혹은 Authentication 등에서 가져올 내용이 굉장히 많아 
추후 사용하지 않을 필드가 너무 많아져 기능 구현의 목적으로만 사용되고 있습니다

좋은 의견 감사합니다. 추후 리팩토링 과정에서 반영하겠습니다.

yooyouny replied:
넵 추후 dto 객체로 분리하면서 요청값의 validation 체크도 따로 해야하지 않을까 싶습니다

yooyouny commented on src/main/java/com/baedalping/delivery/order/service/OrderService.java:
```diff
@@ -0,0 +1,103 @@
+
+    public OrderCreateResponseDto getOrderById(UUID orderId) {
+        Order order = orderRepository.findById(orderId)
+            .orElseThrow(() -> new IllegalArgumentException("Order not found"));
```
예외도 공통된 객체에 넣어 리턴하기로 약속했었는데 이렇게 작성해도 테스트 시 맞게 나왔을까요?
advice가 예외를 잡도록 DeliveryApplicationException으로 만들어 던지든지 아니면 advice에서 핸들러를 따로 만들어야 잡히는거 아닌지 궁금합니다

SerenityZenDev replied:
제가 이슈를 처음사용해서 생긴 문제입니다...ㅠ 
주문 조회에서 사용할 생각으로 만들다가 해당 내용을 제거하는 과정에서 누락된 부분입니다.
주문서 생성단계에서는 사용되지 않는 코드입니다.

다만 질문해주신 내용으로는 advice에 안잡히는게 맞습니다.

yooyouny commented on src/main/java/com/baedalping/delivery/order/service/OrderService.java:
```diff
@@ -0,0 +1,103 @@
+        order.setOrderDetails(orderDetails);
+
+        Order savedOrder = orderRepository.save(order);
+        orderDetailService.saveOrderDetails(orderDetails);
```
OrderService에서 바로 orderDetailRepository를 주입받아 saveAll하지 않고 서비스 객체를 따로 만들어서 거기로 넘기는 이유가 있을까요?

SerenityZenDev replied:
해당 방법도 틀렸다고는 생각하지 않는데 제 생각을 한번 남겨봅니다.

1. 단일책임원칙을 생각했습니다. 
- OrderService와 OrderDetailService 가 각각 하나의 서비스 클래스만 책임을 가지도록 하였습니다.
2. 서비스 계층의 명확한 분리와 테스트 용의함
- 서비스가 독립적으로 실행될 수 있도록 하는게 좋다고 생각했습니다.
- OrderService가 OrderDetailRepository를 직접 사용하면, OrderService를 테스트할 때 OrderDetailRepository와의 의존성을 고려해야 합니다. 반면, OrderDetailService를 통해 접근하면, OrderService의 테스트에서 OrderDetailService를 목(mock)으로 대체할 수 있어 테스트가 더 간단해질것이라 생각했습니다

yooyouny replied:
1. orderService는 주문관련 비즈니스 로직이고 orderDetailService는 주문상세정보를 관리하는 로직으로 서로의 책임에만 집중하는 것이라 좋은것 같습니다.
2. 테스트 역시 각 레이어별로 독립적으로 실행되면 용이할 것 같습니다

리뷰과정은 틀린걸 지적하는게 아니라 코드 작성자의 의도를 파악하는 과정이라고 생각합니다 저 또한 그런 의도로 코멘트를 남기는 것이니 질문과 대답이 자유로웠으면 합니다~_~

yooyouny commented on src/main/java/com/baedalping/delivery/order/entity/Order.java:
```diff
@@ -0,0 +1,69 @@
+    private List<OrderDetail> orderDetails;
+
+//    @Column(name = "order_date", nullable = false)
+//    private LocalDateTime orderDate;
```
혹시 orderDate를 따로 제거한건 audit 필드인 createdAt을 사용하려고 하신걸까요?

SerenityZenDev replied:
여러모로 고민중인 내용입니다만 가장 크게 고민한 이유는 이미 Auditing을 상속받아 사용중인데 order_date를 위해서 Order 클래스에 다시 Auditing을 적용하는 부분도 조금 이상하다는 생각이 들어서 createdAt을 사용하는 것이 좋다고 생각했습니다

yooyouny replied:
비슷한 성격의 필드가 있더라도 의미가 다르면 분리해야한다고 생각해서 적어봤습니다
현재는 간단한 과정이라 createdAt = 주문일자로 정의하고 넘어가도 상관없을것 같습니다 하지만 나중에 확장 했을때 예약주문이 생긴다거나 주문처리지연 등이 있는 경우면 timestamp 값이 조금이라도 다를 수 있으니 orderDate 라는 필드로 따로 관리하는것이 좋을것 같습니다

yooyouny reviewed (commented):
작성하신 부분은 확인 완료했습니다 
dev에 머지한다는 것은 실행시 문제가 없이 해당 로직이 처리된다는걸 확인하는것도 있어서 
테스트 코드를 작성하지 않고 포스트맨으로 API 테스트를 하신거라면 PR 올리실때 스크린샷을 첨부해주시면 좋을것 같아 approve가 아닌 commnet로 드립니다 (성공상황, 예외상황 등의 response 첨부) 
고생하셨습니다

chrysan5 commented on src/main/java/com/baedalping/delivery/order/entity/Order.java:
state 가 enum으로 작성되는 것이므로 @Enumerated 어노테이션을 추가하면 더 좋을 것 같습니다.

chrysan5 commented on src/main/java/com/baedalping/delivery/order/entity/OrderDetail.java:
subtotal의 경우 테이블 명세서에 DECIMAL(10,2)라고 데이터 타입을 지정하셨는데, @Column(precision = 10, scale = 2) 설정과 함께 BigDecimal로 지정하여 사용하는 것은 어떤가요? 제가 아직 pr을 올리지 않았지만 product의 price도 합산을 위해서 BigDecimal로 사용하였습니다. 그러므로 unitPrice도 합의하여 데이터 타입을 같이 맞출 필요할 것으로 보이네요

SerenityZenDev replied:
제가 처음에 몰랐던 내용을 찾아서 변경한 내용인데 공유를 안했네요..

BigDecimal이 불변 객체이기 때문에 정밀도가 높은것은 사실이지만 ORM을 통한 수정의 경우에도 일부 수정이 아닌 객체를 통으로 덮어씌우는 형식이라 성능의 문제도 발생하고 사실 소수점을 이용할 일도 없어서 integer로 변경하였습니다.

chrysan5 replied:
아하 그렇군요 저도 integer로 변경하도록 하겠습니다.

SerenityZenDev committed:
refactor: 주문서 생성 구현 (#2)
- ErrorCode 추가
- Order 엔티티 enum 설정 추가
- 주문서 생성 예외 처리 적용

SerenityZenDev commented:
Postman을 통한 성공 예외처리 이미지 첨부하였습니다. 모두 좋은 의견 감사합니다.

SerenityZenDev committed:
Merge branch 'dev' of https://github.com/BAEDALPING/order-service into feat/2-주문서생성

# Conflicts:
#	src/main/java/com/baedalping/delivery/global/common/exception/ErrorCode.java

SerenityZenDev committed:
Merge branch 'dev' of https://github.com/BAEDALPING/order-service into feat/2-주문서생성

SerenityZenDev committed:
refactor: 주문서 생성 구현 (#2)
- 주문일자 생성
- UUID Validation 추가
- Redis 장바구니를 이용한 주문서 생성 기능 구현
- 주문서 생성을 위한 장바구니 호출
- ErrorCode 추가

SerenityZenDev committed:
chore: 주문서 생성 구현 (#2)
- Cart 패키지 이동
- Order 패키지 이동

yooyouny commented on src/main/java/com/baedalping/delivery/domain/order/dto/OrderCreateRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+@Getter
+public class OrderCreateRequestDto {
+
+    @NotNull(message = "Address ID cannot be null")
```
String 타입이라 빈문자열도 체크할 수 있는 어노테이션이 좋을것 같습니다 패턴의 정규표현식은 UUID 값 형식인지를 확인하는걸까요?

SerenityZenDev replied:
예 맞습니다 해당 패턴이 UUID 값 형식인지를 확인하면서 빈문자열도 함께 체크합니다.

yooyouny commented on src/main/java/com/baedalping/delivery/domain/cart/dto/CartRequestDto.java:
```diff
@@ -1,4 +1,4 @@
-package com.baedalping.delivery.cart.dto;
+package com.baedalping.delivery.domain.cart.dto;
```
어제 이부분을 제대로 못봤던것 같은데 cart의 request에서도 validation 체크가 필요할 것 같습니다 근데 저만의 생각인데 어차피 서비스 내부에서 findByid로 찾아오는 요소들이고 특별히 도메인정책이 있지않는 이상(숫자, 문자로만 구성되어야 한다는 등) dto에서는 널값만 검증해도 되지 않을까 하는 생각입니다

SerenityZenDev replied:
좋은 생각이네요 참고하겠습니다 
cart request validation의 경우 cart 상세 조회 구현 시 같이 진행하겠습니다

yooyouny reviewed (approved):
주문서를 생성하는 로직에서는 별다른 문제가 없는것 같습니다 cart dto부분만 확인해주시면 좋을것같아요
